README.md
# react-grid-debugger

> Visual Tool for debugging grid-based layouts

![Grid Debugger Demo](./grid.gif)

[![NPM](https://img.shields.io/npm/v/react-grid-debugger.svg)](https://www.npmjs.com/package/react-grid-debugger) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

## Table of Contents

- [Motivation](#motivation)
- [Install](#install)
- [Usage](#usage)
- [Props](#props)
- [Should this be a dependency or a devDependency?](#should-this-be-a-dependency-or-a-devdependency)
- [License](#license)


## Motivation
A layout based on a grid system can be something hard to visually debug: proportions, alignments and gutters may vary from project to project and making sure that each layout is well implemented takes both developer and designer time.

This is a visual tool based on a React component that adds a grid overlay on top of our current page, making it easier to see if the number of columns, spacings, etc are up to spec.


## Install

```bash
npm install --save react-grid-debugger
or
yarn add grid-debuger
```

Please check [Should this be a dependency or a devDependency?](#should-this-be-a-dependency-or-a-devdependency) to understand if you need it as a dev dependency or not

## Usage

`react-grid-debugger` is a react component, so you can use it like you'd use any other component.

It accepts a `theme` prop so if your project uses [styled-components](https://github.com/styled-components/styled-components)/[emotion](https://github.com/emotion-js/emotion) and you have defined the `breakpoints` for your theme, you can pass it into the debugger.

```jsx
import React, { Component } from 'react'

import GridDebugger from 'react-grid-debugger'

export default class App extends Component {
  render () {
    return (
      <main>
        <GridDebugger
          show
          maxWidth={'1000px'}
          theme={myAppTheme}
          numCols={[1, 4, 12]}
          gutter={{ 320: '2rem', 680: '40px', 800: '50px'}}
          />
      </main>
    )
  }
}
```

### Toggling the debugger
Pressing keys `Ctrl` and `g` simultaneously toggles the debugger visibility.

### More configuration examples:
```js
<GridDebugger numCols={12} gutter="40px" />
<GridDebugger numCols={12} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} />
<GridDebugger numCols={12} gutter={['2rem', '40px', '50px', '1000px']} />
<GridDebugger numCols={[1, 4, 7]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} />
<GridDebugger maxWidth={'400px'} numCols={[1, 4, 12]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} />
<GridDebugger maxWidth={['280px', '600px', '1000px']} numCols={[1, 4, 12]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} />
```

## Props

Note that most of the props accept:
- `string` or `number`: for indicating a single value for that prop.
- `array`: an array (of numbers or strings) that should be used as values, ordered by breakpoints. This means that the first element of the array will be used for the first breakpoint, the second one for the second breakpoint, and so on. If no `theme` is provided to the debugger, the [default breakpoints](./src/defaultBreakpoints.js) are used. If the number of values provided is bigger than the number of breakpoints, the remaining ones are discarded.
- `object`: an object where each key represents a screen width and its property represents the value that should be used for that breakpoint.

| Prop | Type | Required |Description |
|------|----------|------|-------------|
| show | `bool` | No | wether the debugger should be initially visible (it can be toggled with `ctrl + g`) |
| theme |  No | `object` | represents the app theme, normally provided by a `ThemeProvider`, if you use [styled-components](https://github.com/styled-components/styled-components)/[emotion](https://github.com/emotion-js/emotion) |
| maxWidth | `string / array / object` | No  | grid maximum width |
| numCols | `number / array / object`  | Yes | number of columns |
| gutter | `string / array / object` | Yes | gutter size |


## Should this be a dependency or a devDependency?

It's really up to you. You might want to just have it available locally and not worry abound bundling it, or you may want it to be available on a specific environment. If you do make it available in a prod build, make sure to use a separate chunk for it and only request it if really needed.

## License

MPL-2.0


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.2"></a>
## [1.0.2](https://github.com/yldio/react-grid-debugger/compare/v1.0.1...v1.0.2) (2019-03-23)


### Bug Fixes

* fix readme typo ([b91eb77](https://github.com/yldio/react-grid-debugger/commit/b91eb77))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/yldio/react-grid-debugger/compare/v1.0.0...v1.0.1) (2019-02-12)


### Bug Fixes

* add keywords to package.json ([ea41dbc](https://github.com/yldio/react-grid-debugger/commit/ea41dbc))
* fix table formatting ([9a21445](https://github.com/yldio/react-grid-debugger/commit/9a21445))
* fix typo on readme ([c00cc9a](https://github.com/yldio/react-grid-debugger/commit/c00cc9a))
* improve doc ([1749b94](https://github.com/yldio/react-grid-debugger/commit/1749b94))



<a name="1.0.0"></a>
# 1.0.0 (2019-02-11)


### Bug Fixes

* fix app example prop ([af062d4](https://github.com/yldio/react-grid-debugger/commit/af062d4))
* fix col color ([ee23723](https://github.com/yldio/react-grid-debugger/commit/ee23723))
* remove sideMargins prop - not needed ([bbf1012](https://github.com/yldio/react-grid-debugger/commit/bbf1012))
* rename to react-grid-debugger ([0028efd](https://github.com/yldio/react-grid-debugger/commit/0028efd))
* simplified object calc ([c50a194](https://github.com/yldio/react-grid-debugger/commit/c50a194))


### Features

* add release script ([116e292](https://github.com/yldio/react-grid-debugger/commit/116e292))
* add tests ([e3465e5](https://github.com/yldio/react-grid-debugger/commit/e3465e5))
* added maxwidth ([661ab67](https://github.com/yldio/react-grid-debugger/commit/661ab67))
* added sideMargins prop and key handler ([444eb4b](https://github.com/yldio/react-grid-debugger/commit/444eb4b))
* Documentation and renamed gutters to gutter ([6d16d25](https://github.com/yldio/react-grid-debugger/commit/6d16d25))
* first commit - gutters ([75739c5](https://github.com/yldio/react-grid-debugger/commit/75739c5))
* num cols working ([ecdafe8](https://github.com/yldio/react-grid-debugger/commit/ecdafe8))


src/test.js
import React from 'react'
import renderer from 'react-test-renderer'

import GridDebugger from '../src/index'

it('render null if no `show` prop is passed', () => {
  const tree = renderer.create(
    <GridDebugger 
      maxWidth={'1000px'}
      numCols={12}
      gutter={'4rem'}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
it('renders <GridDebugger /> with single value props correctly', () => {
  const tree = renderer.create(
    <GridDebugger 
      show
      maxWidth={'1000px'}
      numCols={12}
      gutter={'4rem'}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders <GridDebugger /> with object props correctly', () => {
  const tree = renderer.create(
    <GridDebugger 
      show
      maxWidth={{ 320: '300px', 580: 'none', 800: '1000px' }}
      numCols={{ 320: 1, 680: 2, 800: 12 }}
      gutter={{ 320: '2rem', 680: '40px', 800: '50px' }}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders <GridDebugger /> with array props correctly', () => {
  const tree = renderer.create(
    <GridDebugger
      show
      maxWidth={['280px', '600px', '1000px']}
      numCols={[1, 4, 12]}
      gutter={['2rem', '40px', '50px', '1000px']}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});



src/index.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { applyGutters, applyMaxWidth, applyNumCols, applySideMargins } from './rules'
import { getGridProperty } from './utils'

const Grid = styled.div`
  display: grid;
  box-sizing: content-box;
  margin: 0 auto;
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  opacity: 0.6;
  z-index: 1000;
  pointer-events: none;
  grid-template-columns: repeat(${props => props.cols.singleValue}, 1fr);

  /* Side Margins */
  ${props =>
    props.gutter &&
    props.gutter.singleValue &&
    `
    padding: 0 ${props.gutter.singleValue};
  `};

  ${props =>
    props.gutter &&
    props.gutter.map &&
    `
    ${applySideMargins(props.gutter.map)}
  `};

  /* Gutters */
  ${props =>
    props.gutter &&
    props.gutter.singleValue &&
    `
    column-gap: ${props.gutter.singleValue};
  `};

  ${props =>
    props.gutter &&
    props.gutter.map &&
    `
    ${applyGutters(props.gutter.map)}
  `};

  /* Columns */
  ${props =>
    props.cols &&
    props.cols.map &&
    `
    ${applyNumCols(props.cols.map)}
  `};

  /* Max-Width */
  ${props =>
    props.maxWidth &&
    props.maxWidth.singleValue &&
    `
    max-width: ${props.maxWidth.singleValue};
  `};

  ${props =>
    props.maxWidth &&
    props.maxWidth.map &&
    `
    ${applyMaxWidth(props.maxWidth.map)}
  `};
`

const Col = styled.div`
  width: 100%;
  height: 100vh;
  background: #d8fff7;
  border-left: 1px solid #31ffde;
  border-right: 1px solid #31ffde;

  border-left: 1px solid #31ffde;
  border-right: 1px solid #31ffde;
`

export default class GridDebugger extends Component {
  static propTypes = {
    show: PropTypes.bool,
    gutter: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.array,
      PropTypes.string
    ]).isRequired,
    maxWidth: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.array,
      PropTypes.string
    ]),
    numCols: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.object,
      PropTypes.array
    ]).isRequired,
    theme: PropTypes.object
  }

  componentDidMount() {
    window.addEventListener('keyup', this.handleKey)
  }

  componentWillUnmount() {
    window.removeEventListener('keyup', this.handleKey)
  }

  constructor(props) {
    super(props)

    this.state = {
      showGrid: !!props.show
    }
  }

  render() {
    const {
      theme,
      gutter,
      maxWidth,
      numCols
    } = this.props
    const guttersObject = getGridProperty(gutter, this.props.theme)
    const cols = getGridProperty(numCols, theme, 'cols')
    const maxWidthObj = maxWidth
      ? getGridProperty(maxWidth, theme)
      : { singleValue: 'none' }

    return this.state.showGrid ? (
      <Grid
        gutter={guttersObject}
        cols={cols}
        maxWidth={maxWidthObj}
      >
        {Array.from(new Array(cols.singleValue)).map((_, idx) => (
          <Col key={idx} />
        ))}
      </Grid>
    ) : null
  }

  handleKey = ev => {
    if (ev.key === 'g' && ev.ctrlKey) {
      this.setState({
        showGrid: !this.state.showGrid
      })
    }
  }
}


src/rules.js
const buildMediaQueryWithRule = (width, rule, mode = 'min') => {
  return `
    @media (${mode}-width: ${width}px) {
      ${rule}
    }
  `
}

export const applySideMargins = widthMarginMap => {
  return Object.keys(widthMarginMap)
    .map(minWidth => {
      const rule = `padding: 0 ${widthMarginMap[minWidth]};`

      return buildMediaQueryWithRule(minWidth, rule)
    })
    .join('\n')
}

export const applyGutters = widthGutterMap => {
  return Object.keys(widthGutterMap)
    .map(minWidth => {
      const rule = `column-gap: ${widthGutterMap[minWidth]};`

      return buildMediaQueryWithRule(minWidth, rule)
    })
    .join('\n')
}

export const applyNumCols = widthNumColsMap => {
  return Object.keys(widthNumColsMap)
    .map(minWidth => {
      const rule = `grid-template-columns: repeat(${
        widthNumColsMap[minWidth]
      }, 1fr);`

      return buildMediaQueryWithRule(minWidth, rule)
    })
    .join('\n')
}

export const applyMaxWidth = widthMaxWidthMap => {
  return Object.keys(widthMaxWidthMap)
    .map(minWidth => {
      const rule = `max-width: ${widthMaxWidthMap[minWidth]};`

      return buildMediaQueryWithRule(minWidth, rule)
    })
    .join('\n')
}


src/utils.js
import defaultBreakpoints from './defaultBreakpoints'

const getBreakPoints = theme =>
  theme && theme.breakpoints ? theme.breakpoints : defaultBreakpoints

export const getGridProperty = (input, theme, property) => {
  if (typeof input === 'number' || typeof input === 'string') {
    return {
      singleValue: input
    }
  }

  if (typeof input === 'object') {
    if (!Array.isArray(input)) {
      // Input is an object.
      const singleValue =
        property === 'cols'
          ? Math.max(...Object.keys(input).map(width => input[width]))
          : undefined

      return {
        map: input,
        singleValue
      }
    } else {
      // Input is an array.
      const breakpoints = getBreakPoints(theme)
      const breakpointsWidths = Object.keys(breakpoints).map(
        bpName => breakpoints[bpName]
      )
      const map = {}

      breakpointsWidths.forEach((currWidth, idx) => {
        const currItem = input[idx]
        if (currItem != null) {
          map[currWidth] = currItem
        }
      })

      return {
        map,
        singleValue: property === 'cols' ? Math.max(...input) : undefined
      }
    }
  }
}

example/src/App.js
import React, { Component } from 'react'
import { ThemeProvider } from 'emotion-theming'
import GridDebugger from 'react-grid-debugger'

const theme = {
  colors: {
    primary: 'hotpink'
  }
}


export default class App extends Component {
  render () {
    return (
      <ThemeProvider theme={theme}>
      <div>
        {/* <GridDebugger numCols={12} gutter="40px" /> */}
        {/* <GridDebugger numCols={12} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} /> */}
        {/* <GridDebugger numCols={12} gutter={['2rem', '40px', '50px', '1000px']} /> */}
        {/* <GridDebugger numCols={[1, 4, 7]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} /> */}
        {/* <GridDebugger maxWidth={'400px'} numCols={[1, 4, 12]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} /> */}
        {/* <GridDebugger maxWidth={['280px', '600px', '1000px']} numCols={[1, 4, 12]} gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} /> */}
        <GridDebugger 
          show
          maxWidth={'1000px'} 
          numCols={[1, 4, 12]} 
          gutter={{ 320: '2rem', 680: '40px', 800: '50px'}} 
          />
      </div>
      </ThemeProvider>
    )
  }
}
